### PR TITLE
Break up axialmap.h/.cpp

### PR DESCRIPTION
**Pull request by orange-vertex**
This is (I think) the last big change we need to go through to untangle the inheritance madness. The overall aim was to reduce "ShapeGraphs" to a vector<ShapeGraph> within the metagraph to follow along with the pointmaps and datamaps. Unfortunately this one was much more complicated:

[First commit] The ShapeGraphs class essentially contains making all/fewest line maps and the various conversions, but also stores three fields that belong to an all-line map (m_polygons, m_poly_connections, m_radial_lines) used to generate fewest-line maps from the all-line map. Instead of this I span off a new extension of the ShapeGraph called an AllLineMap which contains the three fields as well as the functions for making all/fewest line maps. Fewest line maps are plain ShapeGraphs. Ideally we should now be able to create more All-Line Maps but the current implementation of the .graph files only allows for one (the three fields are stored after all the ShapeGraphs). With this implementation I also pulled out all the other bits and pieces for making fewest line maps, such as the AxialMinimiser the AxialPolygons etc. I also pulled out TidyLines and tolerances in their own files as they are used in a few places. Making all/fewest line maps is tested in the regression tests.

[Second commit] To actually be able to store both ShapeGraphs and AllLineMaps into the same vector I changed it into a vector<unique_ptr<ShapeGraph>> and placed it into the MetaGraph (instead of in the ShapeGraphs class), actually replacing m_shape_maps. For everything to work I also had to pull out the various conversion functions (drawing to axial, segment etc.) into their own file (mapconverter.h/.cpp). For these functions to work outside the ShapeGraphs environment I changed them to take data and return maps instead of them directly appending new maps to the current container. Unfortunately this functionality is not tested yet, hence PR #217 and another to follow along with the regression tests and new binaries before this one is merged.

The last few commits are just to fix various bugs created and eventually completely remove ShapeGraphs as well as its base class ShapeMaps< T > given that this was the last derived class from that base.

Other thoughs:

The conversion functions currently throw RuntimeExceptions to indicate errors in the process but I should probably change this to specialised exceptions (ConversionException?) and catch them in order to show a message.

I should probably change axialmap.h/.cpp to shapegraph.h./cpp

There is some conversion functionality currently in mgraph.cpp (convertToDrawing) which should probably go to the converter as well.

Maybe I should move the conversion choices in the MapConverter and let the MetaGraph just send the target map type wanted as in PR #217 

The number of files in salalib is growing and growing... Maybe we need to think of a folder hierarchy

### CODE REVIEW

File: salalib/axialpolygons.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+   bool segend;
+   // padding the remaining three bytes behind the bool - don't use int : 24 as this will grab the next 4 byte block
+   char pad1 : 8;
+   short pad2 : 16;
+
+   RadialKey(const AxialVertexKey& v = NoVertex, float a = -1.0f, bool se = false) : pad1(0), pad2(0)
+   { vertex = v; ang = a; segend = se; }
+   RadialKey(const RadialKey& rk) : pad1(0), pad2(0)
+   { vertex = rk.vertex; ang = rk.ang; segend = rk.segend; }
+   friend bool operator < (const RadialKey& a, const RadialKey& b);
```
Comment (blsemo) [inline salalib/axialpolygons.h]: Why do they need to be friends? Nothing private here...

File: salalib/mapconverter.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include "salalib/axialmap.h"
+#include "salalib/shapemap.h"
+#include "salalib/spacepixfile.h"
+#include "genlib/comm.h"
+
+namespace MapConverter {
```
Comment (blsemo) [inline salalib/mapconverter.h]: I like it that these are just functions  that do something to a map and return the result. Ideally, could we do something similar for the AllLines Map?

File: salalib/mgraph.cpp [hunk 3]
```diff
@@ -1133,10 +1174,50 @@ bool MetaGraph::makeFewestLineMap( Communicator *communicator, int replace )
...
+   bool retvar = true;
 
    try {
-      retvar = m_shape_graphs.makeFewestLineMap(communicator, (replace != 0));
+       // no all line map
+       if (m_all_line_map == -1) {
+          return false;
+       }
+
+       AllLineMap* alllinemap = dynamic_cast<AllLineMap*>(m_shapeGraphs[size_t(m_all_line_map)].get());
```
Comment (blsemo) [inline salalib/mgraph.cpp]: Hmm, not sure I like this.
Isn't an all line map just a shape graph of which you know it's an all line map?
It seems we can have one all line map at max (as we only have on index to it), so why not have that as a special entity, or have the extra data somewhere else?
At least, the result of a `dynamic_cast` needs to be checked for null and the error needs to be clear...

File: salalib/tidylines.h [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include "salalib/spacepix.h"
+
+// helpers... a class to tidy up ugly maps people may give me...
+
+class TidyLines : public SpacePixel
+{
+public:
+   TidyLines() {;}
```
Comment (blsemo) [inline salalib/tidylines.h]: I am not sure I like the shape of this class. If I understand correctly, you have to construct it with the standard constructor, and then feed the data via the `tidy` or `quicktidy` methods which will turn this into a valid SpacePixel.
Wouldn't it be better if we had two classes that took the data in the constructor and constructed a tidy lines SpacePixel on construction?

File: salalib/tidylines.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+   virtual ~TidyLines() {;}
```
Comment (blsemo) [inline salalib/tidylines.h]: Is some other class derived from `TidyLines`? Or why do we need an empty virtual destructor?

File: salalib/alllinemap.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include "salalib/axialmap.h"
+#include "salalib/axialpolygons.h"
+
+class AllLineMap: public ShapeGraph
+{
+public:
+    AllLineMap(Communicator *comm,
```
Comment (blsemo) [inline salalib/alllinemap.h]: I don't particularly like the structure of this all line map - it does three things in a weird mix:
- on construction, it actually calculates an All Line Map. 
- it has some extra data that is required for an all line map, but that is public and can be accessed for read and write from external. I am assuming that the constructor that just takes a name is used to create an empty all line map and fill it with data from externally when de-serialising?
- it provides methods to reduce an all line map to a fewest line map.

I am wondering whether we wouldn't be better off with a data structure that just holds data, and functions to do the calculations?

**Comment by orange-vertex** [comment]
Merging this as it's a pretty big base change and will address the All-Line map issues in another PR

**Pull request merged**